File: src/main/java/eu/qwsome/sql/condition/Exists.java
package eu.qwsome.sql.condition;


import eu.qwsome.sql.Select;

import java.util.Arrays;

import static eu.qwsome.sql.SubselectValueHolder.subselect;

public class Exists implements Condition {

  private final ValueHolder[] values;


  Exists(final ValueHolder... values) {
    this.values = values;
  }


  public static Exists exists(final ValueHolder values) {
    return new Exists(values);
  }


  public static Exists exists(final Select.ConditionsBuiltPhase subselect) {
    return new Exists(subselect(subselect));
  }


  @Override
  public CharSequence get() {
    final var builder = new StringBuilder();
    appendTo(builder);

    return builder;
  }


  @Override
  public ValueConstructor getValues() {
    final var valueConstructor = new ValueConstructor();
    Arrays.stream(this.values).forEach(valueConstructor::add);
    return valueConstructor;
  }


  @Override
  public void appendTo(final StringBuilder builder) {
    builder
      .append("EXISTS ");

    int i = 0;
    for (; i < this.values.length - 1; i++) {
      builder.append(this.values[i].getSql())
        .append(", ");
    }

    builder.append(this.values[i].getSql());
  }
}

File: src/main/kotlin/eu/qwsome/sql/condition/Exists.kt
/*
 * @author Lukáš Kvídera
 */
package eu.qwsome.sql.condition

import eu.qwsome.sql.Select
import eu.qwsome.sql.SubselectValueHolder

class Exists private constructor(private vararg val values: ValueHolder) : Condition {

    override fun getValues(): ValueConstructor {
        val valueConstructor = ValueConstructor()
        values.forEach { valueConstructor.add(it) }
        return valueConstructor
    }

    override fun appendTo(builder: StringBuilder) {
        builder.append("EXISTS ")

        values.forEachIndexed { index, value ->
            if (index > 0) {
                builder.append(", ")
            }
            builder.append(value.getSql())
        }
    }

    companion object {
        @JvmStatic
        fun exists(values: ValueHolder) = Exists(values)

        @JvmStatic
        fun exists(subselect: Select.ConditionsBuiltPhase) = Exists(SubselectValueHolder.subselect(subselect))
    }
}
